### PR TITLE
Assignment 7

### PR DESCRIPTION
**Pull request by ErikCLabrot**
Assignment 7 
-EKF
-EKF Node
-ICP Improvements
-ICP Node Improvements
-KDTree Implementation

### CODE REVIEW

File: applied_robotics/applied_robotics/ekf_node.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        self.node_name = "ekf_node"
+        super().__init__(self.node_name)
+
+        #Init Params
+        self.lidar_odometry_topic = self.declare_parameter('lidar_odometry_topic', '/lidar/odometry').value
+        self.ekf_state_topic = self.declare_parameter('ekf_state_topic', '/filtered_odometry').value
+        self.imu_sensor_topic = self.declare_parameter('imu_sensor_topic', '/robot/imu').value
+        self.imu_sensor_update_rate = self.declare_parameter('imu_sensor_update_rate', 1000).value
+        self.start_position = self.declare_parameter('world_start_position', [0.0, 0.0, 0.0]).value
+        self.start_orientation = self.declare_parameter('world_start_orientation', 0.0).value
```
Comment (slensgra) [inline applied_robotics/applied_robotics/ekf_node.py:58]: Nice! Good use of the parameters

File: applied_robotics/applied_robotics/ekf_node.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        '''
+        IMU sensor callback, EKF predict step. Predict based on latest IMU reading whenever triggered.
+
+        args:
+            msg: IMU data from ROS
+        '''
+        imu_angular_velocity = msg.angular_velocity.z
+        imu_linear_acceleration = msg.linear_acceleration.x
+        
+        self.ekf_state_estimator.predict(imu_angular_rate = imu_angular_velocity, imu_linear_acceleration = imu_linear_acceleration)
```
Comment (slensgra) [inline applied_robotics/applied_robotics/ekf_node.py:126]: I think it is typical not to have the space in the equal sign when using keyword arguments

File: applied_robotics/applied_robotics/extended_kalman_filter.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+Date: 4/8/2025
+
+This module defines the implementation for an extended kalman filter that uses an IMU to
+predict based off its sensing of control inputs, and updates based on velocity received from
+a lidar odometry module.
+"""
+
+import numpy as np
+
+class ekf():
```
Comment (slensgra) [inline applied_robotics/applied_robotics/extended_kalman_filter.py:14]: Class names should be upper camel case, so Ekf() instead

File: applied_robotics/applied_robotics/icp_node.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+        args:
+            scan: laserscan from lidar received through ROS subscription
+        '''
+
+        scan_time = scan.header.stamp
+        current_compass = self.current_compass_heading
+
+        if self.last_scan is not None:
+            Transform = self.icp.compute_registration(source_scan = scan,
```
Comment (slensgra) [inline applied_robotics/applied_robotics/icp_node.py:236]: Since it is typical that a class will be in upper camel case, having the variable called Transform makes this read like the compute registration is returning the specification of a class or something like that.

File: applied_robotics/applied_robotics/pose_plot_node.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+from rclpy.node import Node
+from tf2_msgs.msg import TFMessage
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import Pose, Point, Twist, Quaternion, TransformStamped
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+from matplotlib import style
+
+
+class PosePlotNode(Node):
```
Comment (slensgra) [inline applied_robotics/applied_robotics/pose_plot_node.py:24]: The node for plotting a pose is great -- the file is renamed from last assignment I assume?

File: applied_robotics/applied_robotics/lidar_variance_measurement.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    def compute_variance(self):
+        '''
+        Function to write mean and variance of data to file
+        '''
+        linear_array = np.array(self.linear_velocities)
+        angular_array = np.array(self.angular_velocities)
+
+        linear_mean = linear_array.mean()
+        linear_variance = linear_array.var()
```
Comment (slensgra) [inline applied_robotics/applied_robotics/lidar_variance_measurement.py:71]: This works, but it might be more helpful to look at the covariance matrix of the velocity measurements since it is a 2d measurement (x, y). That would give something that would more easily translate to what you need with the EKF. You can compute the covariance of a few samples by looking at this numpy function: https://numpy.org/doc/stable/reference/generated/numpy.cov.html

File: applied_robotics/applied_robotics/kd_tree.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+This module defines the implementation for a KDTree added onto the lidar odometry node.
+In theory, this speeds the nearest neighbor search to O(n log n) from O(n^2) per wiki. 
+In practice, python prevents this from working expediently due to the reliance on recursive calls.
+"""
+
+
+import numpy as np
+import sys
+
+class KDNode():
```
Comment (slensgra) [inline applied_robotics/applied_robotics/kd_tree.py:16]: It is awesome that you implemented this yourself and good point that in python, fancy data structures tend to be slower than more plain implementations. Where using a better datastructure often helps is when you can take advantage of an existing library that is maybe wrapping a faster c++ implementation. This is what pandas and numpy do to make things fast.

An example kd-tree that might actually improve things is here: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KDTree.html

File: applied_robotics/applied_robotics/extended_kalman_filter.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        linear_velocity = self.state_vector[3,0]
+
+        linear_velocity_new = linear_velocity + imu_linear_acceleration * self.dt
+
+        #Predict State (state_k+1 = state_k + dynamics)
+        self.state_vector[0,0] += linear_velocity * np.cos(theta) * self.dt
+        self.state_vector[1,0] += linear_velocity * np.sin(theta) * self.dt
+        self.state_vector[2,0] += imu_angular_rate * self.dt
+        self.state_vector[3,0] = linear_velocity_new
+        self.state_vector[4,0] = imu_angular_rate
```
Comment (slensgra) [inline applied_robotics/applied_robotics/extended_kalman_filter.py:95]: Phrasing this as a matrix multiply and some constants could make this a little more clear to read

File: applied_robotics/applied_robotics/extended_kalman_filter.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            [0, 1, linear_velocity * np.cos(theta) * self.dt, np.sin(theta) * self.dt, 0],
+            [0, 0, 1, 0, 0],
+            [0, 0, 0, 1, 0],
+            [0, 0, 0, 0, 1]
+        ])
+
+        #Predict Covariance Matrix
+        self.state_covariance = jacobian @ self.state_covariance @ jacobian.T + self.process_noise_covariance
+
+    def update(self, measurement_vector):
```
Comment (slensgra) [inline applied_robotics/applied_robotics/extended_kalman_filter.py:111]: Great!

File: applied_robotics/applied_robotics/icp.py [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+        returns:
+            points: vertically stacked x,y pairs as np array
+        '''
+        x = ranges * np.cos(angles)
+        y = ranges * np.sin(angles)
+        points = np.vstack((x,y)).T
+
+        return points
+
+    def undistort_laser_scan(self, scan, linear_velocity, angular_velocity, lidar_rate=10):
```
Comment (slensgra) [inline applied_robotics/applied_robotics/icp.py:85]: Great work on researching this and adding it in

**Review by slensgra — APPROVED** [review]
For future assignments, make sure you merge in your prior PR so it is easy to tell what is new. Other than a couple of naming tweaks, this looks really awesome
